plenoptic/simulate/models/frontend.py

```diff
@@ -1,96 +1,407 @@
-        filters will still be 31x31. Major changes in the future will allow users to specify kernel size.
-    requires_grad: bool
-        Whether or not model is trainable.
+    size:
+        Filter kernel size.
+    std:
+        Standard deviation of 2D circular Gaussian.
+
     Returns
     -------
-    y: torch.Tensor
-        representation (B, 2, H, W)
+    filt: Tensor
+    """
+    assert std > 0.0, "stdev must be positive"
+
+    device = std.device
+
+    if isinstance(size, int):
+        size = (size, size)
+
+    origin = torch.tensor(((size[0] + 1) / 2.0, (size[1] + 1) / 2.0), device=device)
+
+    shift_y = torch.arange(1, size[1] + 1, device=device) - origin[1]
+    shift_x = torch.arange(1, size[0] + 1, device=device) - origin[0]
+
+    (xramp, yramp) = torch.meshgrid(shift_y, shift_x)
+
+    log_filt = ((xramp ** 2) + (yramp ** 2)) / (-2.0 * std ** 2)
+
+    filt = torch.exp(log_filt)
+    filt = filt / filt.sum()  # normalize
+
+    return filt
+
+
+class Gaussian(nn.Module):
+    """Isotropic Gaussian single-channel convolutional filter.
+    Kernel elements are normalized and sum to one.
+
+    Parameters
+    ----------
+    kernel_size:
+        Size of convolutional kernel.
+    std:
+        Standard deviation of circularly symmtric Gaussian kernel.
+    pad_mode:
+        Padding mode argument to pass to `torch.nn.functional.pad`.
+    """
+
+    def __init__(
+        self,
+        kernel_size: Union[int, Tuple[int, int]],
+        std: float = 3.0,
+        pad_mode: str = "circular",
+    ):
+        super().__init__()
+        assert std > 0, "Gaussian standard deviation must be positive"
+        self.std = nn.Parameter(torch.tensor(std))
+        self.kernel_size = kernel_size
+        self.pad_mode = pad_mode
+
+    def get_filter(self):
+        filt = circular_gaussian(self.kernel_size, self.std)
+        return filt.view(1, 1, *filt.shape)
+
+    def forward(self, x: Tensor) -> Tensor:
+        self.std.data = self.std.data.abs()  # ensure stdev is positive
+        filt = self.get_filter()
+
+        x = same_padding(x, self.kernel_size, pad_mode=self.pad_mode)
+        y = F.conv2d(x, filt)
+        return y
+
+
+class CenterSurround(nn.Module):
+    """Center-Surround, Difference of Gaussians (DoG) filter model. Can be either
+    on-center/off-surround, or vice versa.
+
+    Filter is constructed as:
+    .. math::
+        f &= amplitude_ratio * center - surround \\
+        f &= f/f.sum()
+
+    The signs of center and surround are determined by `center` argument. The standard
+    deviation of the surround Gaussian is constrained to be at least `width_ratio_limit`
+    times that of the center Gaussian.
+
+    Parameters
+    ----------
+    center:
+        Dictates whether center is on or off. The surround always will be the opposite
+        of the center. Must be either ['on', 'off']; default is 'on' center.
+    kernel_size: Union[int, Tuple[int, int]], optional
+    width_ratio_limit:
```

Comment:
Having `center_std` as well as `surround_std` gives 2 DOF in the model. `width_ratio_limit` just adds constraint to `surround_std`.

If `width_ratio_limit=2`, then `center_std=1` and `surround_std=2` is valid, but `surround_std=2.01` is also valid.

Reply:
`width_ratio_limit` is a lower bound rather than an upper bound. The Surround Gaussian can't be too skinny, and `width_ratio_limit` sets a bound on how skinny it is allowed to be, relative to the Center Gaussian.
I'll make this more clear int he docs.